From: Werner Koch <wk@gnupg.org>
Date: Thu, 19 Oct 2017 17:05:39 +0200
Subject: gpg: Make --dry-run work for secret keys.

* g10/import.c (import_secret_one): Check for dry-run before
transferring keys.
--

The use of --dry-run had no effect when
importing a secret key and the public key already existed.  If the
public key did not exist an error message inhibited the import of the
secret key.

(cherry picked by dkg from commit 68c8619114fd5f24cb6bfb9e0f25c428a8805323)
Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>

Note that we ignore "--import-options show-only" the IMPORT_DRY_RUN
flag, since that feature is not available in stretch. (only --dry-run)
---
 g10/import.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/g10/import.c b/g10/import.c
index 2c748823e..bddb2cfc7 100644
--- a/g10/import.c
+++ b/g10/import.c
@@ -2278,7 +2278,8 @@ import_secret_one (ctrl_t ctrl, kbnode_t keyblock,
       /* At least we cancel the secret key import when the public key
 	 import was skipped due to MERGE_ONLY option and a new
 	 key.  */
-      if (stats->skipped_new_keys <= nr_prev)
+      if (!opt.dry_run
+          && stats->skipped_new_keys <= nr_prev)
 	{
           /* Read the keyblock again to get the effects of a merge.  */
           /* Fixme: we should do this based on the fingerprint or
